modules/OCPP/OCPP.cpp

```diff
@@ -57,6 +62,72 @@ void OCPP::init() {
             return false;
         }
     });
+    this->charge_point->register_upload_diagnostics_callback([this](std::string location) {
+        // create temporary file
+        std::string date_time = "DATETIME"; // FIXME
+        std::string file_name = "diagnostics-" + date_time + "-%%%%-%%%%-%%%%-%%%%";
+
+        auto diagnostics_file_name = boost::filesystem::unique_path(boost::filesystem::path(file_name));
+        auto diagnostics_file_path = boost::filesystem::temp_directory_path() / diagnostics_file_name;
+
+        auto diagnostics = json::object({{"diagnostics", json::object({{"key", "value"}})}});
+        std::ofstream diagnostics_file(diagnostics_file_path.c_str());
+        diagnostics_file << diagnostics.dump();
+        this->upload_diagnostics_thread = std::thread([this, location, diagnostics_file_name, diagnostics_file_path]() {
+            auto diagnostics_uploader = boost::filesystem::path("bin/diagnostics_uploader.sh");
+
+            boost::process::ipstream stream;
+            std::vector<std::string> args = {location, diagnostics_file_name.string(), diagnostics_file_path.string()};
+            boost::process::child cmd(diagnostics_uploader, boost::process::args(args),
+                                      boost::process::std_out > stream);
+            std::string temp;
+            while (std::getline(stream, temp)) {
+                if (temp == "uploaded") {
+                    this->charge_point->send_diagnostic_status_notification(ocpp1_6::DiagnosticsStatus::Uploaded);
+                } else if (temp == "failed") {
+                    this->charge_point->send_diagnostic_status_notification(ocpp1_6::DiagnosticsStatus::UploadFailed);
+                } else {
+                    this->charge_point->send_diagnostic_status_notification(ocpp1_6::DiagnosticsStatus::Uploading);
+                }
+            }
+            cmd.wait();
+        });
+
+        return diagnostics_file_name.string();
+    });
+    this->charge_point->register_update_firmware_callback([this](std::string location) {
+        // // create temporary file
+        std::string date_time = "DATETIME"; // FIXME
+        std::string file_name = "firmware-" + date_time + "-%%%%-%%%%-%%%%-%%%%";
+
+        auto firmware_file_name = boost::filesystem::unique_path(boost::filesystem::path(file_name));
+        auto firmware_file_path = boost::filesystem::temp_directory_path() / firmware_file_name;
+
+        this->update_firmware_thread = std::thread([this, location, firmware_file_path]() {
```

Comment:
there's no detach or join for the threads, this might lead to hard to understand exits in case of exceptions

Reply:
We should add the detach here before merging

Reply:
done